runtime:netrw: simplify gx file handling

Konfekt opened the pull request:
Take care of gx in Git Bash and WSL as maintenance stalled.

The whole logic was somewhat convoluted with some parts repeatedly invoking failed commands.

The file handling was outdated, for example, nowadays Netscape is rarely used, and also opinionated, for example mainly Microsoft Paint and Gimp for Image files.
Instead, let (xdg-)open/start handle this and respect the user's preferences.

Uses a public `:Launch/Open` command to launch applications / open URLs and files similar to Neovim's native [vim.ui.open()](https://github.com/neovim/neovim/commit/af6e6ccf3dee815850639ec5613dda3442caa7d6)

Konfekt committed:
runtime:netrw: simplify gx file handling

It did not work very well, at least on Debian 12, and I am not sure Git
Bash and WSL, for example, were taken care of as maintenance stalled.

The whole logic was somewhat convoluted with some parts repeatedly invoking
failed commands.

The file handling was outdated, for example, nowadays Netscape is rarely
used, and also opinionated, for example mainly Microsoft Paint and Gimp for
Image files.
Instead, (xdg-)open respects the user's preferences.

saccarosium commented on runtime/autoload/netrw.vim:
```diff
@@ -5570,12 +5469,52 @@ fun! netrw#GX()
+
+  return ''
+endf
+
```
I don't think it is a good idea to support specifically markdown and html urls. I think keeping the current behavior is better (only working if the cursor is on an actuall url)

Konfekt replied:
Yes, a matter of taste. I deemed markdown links common enough, but these can be deferred to a suitable ftplugin

Konfekt replied:
Mind you that you can already override the selection to be opened, and this option is preserved

saccarosium replied:
> Mind you that you can already override the selection to be opened, and this option is preserved

What do you mean?

Konfekt replied:
I meant `:help g:netrw_gx` though I find it rather coarse and in general maybe some options are no longer that useful. I dared to remove `g:netrw_browsex_viewer` as it relied by default on ancient applications such as Netscape and seemed overengineered for its user base (that I suspected to be small since by now somebody would have replaced Netscape Navigator by something more modern).

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,67 @@ set cpo&vim
+if has('unix')
+  if has('win32unix') " Git Bash provides /usr/bin/start script calling cmd.exe //c
+    " use start //b "" to set void title and avoid ambiguity with passed argument
+    silent! command -complete=shellcmd -nargs=1 -bang Launch
```
I don't think this is a good user command. It would be probably be better as a function in `autoload/dist`. I don't know if the users wants to launch a program directly.

Konfekt replied:
See https://github.com/vim/vim/issues/14872 for examples; to read any output of input in Vim, say binary, or even text, like a web/html page. One could also imagine a background process like browser-sync for web developers, or more generally a file watcher

saccarosium replied:
Yes, I do think that this functionality is useful. But I find difficult imagine a situation where a user wants the command handy to launch an app. I think this is more useful as a function when you are scripting.

Konfekt replied:
I think this is current practice. At the same time, there are so many plug-ins whose principal raison d'etre is launching some command, say https://github.com/romainl/vim-devdocs, https://github.com/wolandark/vim-live-server/, various markdown live preview [plug-ins](https://github.com/arminveres/md-pdf.nvim/issues/19) vs. launching [Okular](https://gist.github.com/Konfekt/c2066abc31b41430417310bd6e6d795e). Similar to using `:compiler` instead of installing linter or compiler plug-ins, one could opt for `:Launch` instead of installing application specific plug-ins.

chrisbra replied:
Why the `:silent!`?

Konfekt replied:
Thank you. Since there's the `g:loaded_netrwPlugin` load guard, I'd say it's dispensable and dropped it.

saccarosium reviewed (commented):
I've tested this branch and I found that it tries to open the program directly in the CWD.

Step to repro:
1. `sudo make install`
2. `vim`
3. `:Launch <program>`

I get the following error:
```
nohup: failed to run command '/home/sacca/Repos/github.com/vim/vim/xdg-open /home/sacca/Repos/github.com/vim/vim/README.rux.txt': No such file or directory
```

k-takata commented on runtime/doc/pi_netrw.txt:
```diff
@@ -1,29 +1,29 @@
-	    ------------------------------------------------
+            ------------------------------------------------
+            NETRW REFERENCE MANUAL    by Charles E. Campbell
+            ------------------------------------------------
```
Did you intentionally expand the tabs?

Konfekt replied:
No, I let git diff suppress them and sometimes stumble over it. Today is the day

k-takata replied:
The tabs in all the other parts have also been expanded. So, this file has conflicts.

k-takata commented on runtime/doc/pi_netrw.txt:
```diff
@@ -1469,119 +1469,79 @@ With either form of the command, netrw will first ask for confirmation
-	* when editing, hit gx with the cursor atop the special filename
-<	  (latter not available if the |g:netrw_nogx| variable exists)
+        * hitting gx with the cursor atop the file path or alternatively x
+          in in a netrw buffer; the former can be disabled by defining the
```
```suggestion
          in a netrw buffer; the former can be disabled by defining the
```

k-takata commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,100 @@ set cpo&vim
+
+" set up redirection (avoids browser messages)
+" by default, g:netrw_suppress_gx_mesg is true
+if get (g:, ':netrw_suppress_gx_mesg', 1)
```
```suggestion
if get(g:, ':netrw_suppress_gx_mesg', 1)
```

k-takata commented on runtime/autoload/netrw.vim:
```diff
@@ -24,8 +24,8 @@
 "   2024 Aug 22 by Vim Project: fix mf-selection highlight (#15551)
 "   2024 Aug 22 by Vim Project: adjust echo output of mx command (#15550)
 "   2024 Sep 15 by Vim Project: more strict confirmation dialog (#15680)
-"   2024 Sep 19 by Vim Project: mf-selection highlight uses wrong pattern (#15700)
```
Mistake? (Or, needs to be rebased?)

Konfekt replied:
Somehow got lost in between, readded. Thank you for your input!

Konfekt committed:
remove superfluous silent!'s in view of load guard

See [0]

___

[0]: https://github.com/vim/vim/pull/15721#discussion_r1809502879

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,100 @@ set cpo&vim
+    if has("win32")
+      let s:redir= substitute(&srr,"%s","nul","")
+    else
+      let s:redir= substitute(&srr,"%s","/dev/null","")
```
Here you can use `printf` like so:
```vim
printf(&srr, null)
```

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,100 @@ set cpo&vim
+  let g:netrw_regex_url = '\%(\%(http\|ftp\|irc\)s\?\|file\)://\S\{-}'
+endif
+
+silent! function Netrw_get_URL_markdown()
```
Can this be removed? I think this should be remain general. The user can always just go over the url.

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,100 @@ set cpo&vim
+  endtry
+endfunction
+
+silent! function Netrw_get_URL_html()
```
Can this be removed? I think this should be remain general. The user can always just go over the url.

Konfekt replied:
I made them opt-in

Konfekt replied:
Going over the URL is not so easy if concealing is used, for example.

saccarosium reviewed (commented):
I think this can be simplified a little bit in the coming days I will try to give more feedback. Thank you for your patience.

Konfekt committed:
use more modern printf() instead of legacy substitute()

Konfekt committed:
remove file type specific handlers and propose in doc

Konfekt committed:
add note on concealed (markdown) links

k-takata commented on runtime/doc/pi_netrw.txt:
```diff
@@ -1469,42 +1469,82 @@ With either form of the command, netrw will first ask for confirmation
+
+    * for Windows                     : explorer.exe is used
   * for Mac OS X               : open is used.
+    * for Linux               : xdg-open is used.
```
not well aligned.

Konfekt replied:
Thank you, corrected

k-takata commented on runtime/doc/pi_netrw.txt:
```diff
@@ -2533,7 +2573,7 @@ your browsing preferences.  (see also: |netrw-settings|)
 					"kfmclient exec"
 					"gnome-open"
-<
+
```
It seems `<` is needed.

Konfekt replied:
True, accidentally removed as help files conceal by default these

k-takata commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,72 @@ set cpo&vim
+    if has("win32")
+      let s:redir= printf(&srr, null)
+    else
+      let s:redir= printf(&srr, null)
```
```suggestion
      let s:redir= printf(&srr, "nul")
    else
      let s:redir= printf(&srr, "/dev/null")
```

Konfekt replied:
Dang, thank you!

Konfekt committed:
apply kind suggestions by @k-takata

k-takata commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,72 @@ set cpo&vim
+if get(g:, ':netrw_suppress_gx_mesg', 1)
+  if &srr =~# "%s"
+    if has("win32")
+      let s:redir= printf(&srr, "null")
```
```suggestion
      let s:redir= printf(&srr, "nul")
```

Konfekt committed:
fix typo kindly pointed out by @k-takata

Co-authored-by: K.Takata <kentkt@csc.jp>

chrisbra commented on runtime/autoload/netrw.vim:
```diff
@@ -5563,12 +5471,36 @@ fun! netrw#GX()
+  let file = expand("<cfile>")
+  if filereadable(file) | return file | endif
+  " ... or in that of the current buffer?
+  let path = expand('%')
```
```suggestion
  let path = fnamemodify(expand('%'), ':p')
```

chrisbra replied:
and then later on: `let dir = path`

mgedmin replied:
Why `fnamemodify(expand('%'), ':p')` instead of just `expand('%:p')`?

Konfekt replied:
Good point, seems the same

chrisbra replied:
I prefer the former over the latter, because the intention is more clear and therefore not so obscure

chrisbra commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,72 @@ set cpo&vim
+if get(g:, ':netrw_suppress_gx_mesg', 1)
+  if &srr =~# "%s"
+    if has("win32")
+      let s:redir= printf(&srr, "nul")
```
```suggestion
      let s:redir= printf(&srr, has("win32") ? "nul" : "/dev/null")
```

chrisbra commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,72 @@ set cpo&vim
+else
+  let s:redir= ""
+endif
+"  call Decho("set up redirection: redir{".redir."} srr{".&srr."}",'~'.expand("<slnum>"))
```
remove this commented line please

chrisbra commented on runtime/autoload/netrw.vim:
```diff
@@ -5563,12 +5471,36 @@ fun! netrw#GX()
+   endif
+   if !empty(URL) | return URL | endif
+  " URLs end in letter, digit or forward slash
+  let URL = matchstr(expand("<cWORD>"), '\<' .. g:netrw_regex_url .. '\ze[^A-Za-z0-9/]*$')
```
`g:netrw_regex_url` uses `\S`, so that `\ze[^A-Za-z0-9/]*` seems redundant?

Konfekt replied:
Seemingly, though uses non-greedy `{-}` so that we assume that a URL ends in an alphanumeric letter or forward slash.

Konfekt replied:
This keeps it working on URLs with trailing punctuation, parentheses, ...

chrisbra replied:
So shouldn't be anything up to the first white-space then considered an URL? so something like this: `[^ ]*` ?

Konfekt replied:
Likely a misunderstanding, but the idea is that, say, the trailing comma `,` in `.. to https://vim.org, but ...` is not part of the URL `https://vim.org`.

chrisbra replied:
I see. sounds fragile, but we can keep if for now.

Konfekt replied:
Maybe in theory not completely waterproof but [in practice](https://github.com/qadzek/link.vim/issues/12#issuecomment-2343590178)

chrisbra commented on runtime/autoload/netrw.vim:
```diff
@@ -5447,86 +5428,13 @@ fun! netrw#BrowseX(fname,remote)
-"   call Decho("(netrw#BrowseX) macunix and open",'~'.expand("<slnum>"))
-   call s:NetrwExe("sil !open ".s:ShellEscape(fname,1)." ".redir)
-   let ret= v:shell_error
+    exe 'Launch' viewer viewopt escape(fname, '#%')
```
```suggestion
    exe 'Launch' viewer viewopt shellescape(fname,1)
```

chrisbra commented on runtime/autoload/netrw.vim:
```diff
@@ -5447,86 +5428,13 @@ fun! netrw#BrowseX(fname,remote)
-
-  if ret
-   call netrw#ErrorMsg(s:ERROR, "Failed to open '".a:fname."'", 0)
+    exe 'Open' escape(fname, '#%')
```
same here

Konfekt replied:
For user's convenience, I opted to shellescape directly in `:Open` so that  command-line completed paths open

Konfekt committed:
apply @chrisbra 's kind comments [0]

___

[0]: https://github.com/vim/vim/pull/15721#pullrequestreview-2390564538

Konfekt committed:
ensure variable always set

chrisbra commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,63 @@ set cpo&vim
+  if has('win32unix') " Git Bash provides /usr/bin/start script calling cmd.exe //c
+    " use start //b "" to set void title and avoid ambiguity with passed argument
+    command -complete=shellcmd -nargs=1 -bang Launch
+          \ exe 'silent ! start "" //b ' . trim(<q-args>)  s:redir | redraw!
```
is this `//b` correct?

Konfekt replied:
[I dug](https://github.com/junegunn/fzf/pull/3798/commits) into the [MSYS2 Path conversion magic docs](https://www.msys2.org/docs/filesystem-paths/) to ensure that `/b` is not misinterpreted as a path `\b`.
However, this assumes the `start` script coming with MSYS2 / Git Bash; I am not sure about Cygwin and maybe an additional case distinction is needed (as it only knows `cygstart`!?). I assumed it essentially superseded by MSYS2, Git Bash and WSL and would await feedback.

chrisbra replied:
okay fine. I think you may want to adjust the comment above however. Also the comment mentions `start //b ""` while the code uses `start "" //b`

chrisbra commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,63 @@ set cpo&vim
+    command -complete=shellcmd -nargs=1 -bang Launch execute ':silent !'..
+          \ ((<q-args> =~? '\v<\f+\.(exe|com|bat|cmd)>') ?
+            \ 'cmd.exe /c start "" /b ' . trim(<q-args>) :
+            \ 'nohup ' trim(<q-args>) s:redir '&')
```
this uses a weird mix of `..`, `.` or no concatenation operators

chrisbra commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,63 @@ set cpo&vim
+        \ exe 'silent !'.. (&shell =~? '\<cmd\.exe\>' ? '' : 'cmd.exe /c')
+        \ 'start /b ' trim(<q-args>) s:redir | redraw!
+endif
+" if exists(':Launch') == 2
```
Can you enable this condition please?

chrisbra commented on runtime/doc/pi_netrw.txt:
```diff
@@ -1469,42 +1469,82 @@ With either form of the command, netrw will first ask for confirmation
+    * for Mac OS X               : open is used.
+    * for Linux               : xdg-open is used.
+
+To launch a specific application <app>, type :Launch <app>.
```
```suggestion
To launch a specific application <app>, type :Launch <app> <filename>
```

Konfekt replied:
I forgot about `:Open`. Launch is the general background app launcher. It's all documented now.

chrisbra reviewed (commented):
almost there, just a few minor things I noticed.

Konfekt committed:
apply @chrisbra 's kind suggestions [0]

___

[0]: https://github.com/vim/vim/pull/15721#pullrequestreview-2393497092

Konfekt committed:
clarify MSYS2 path conversion in comment [0]

___

[0]: https://github.com/vim/vim/pull/15721#discussion_r1815593815

Konfekt committed:
doc: discuss need for user escaping in :Open/Launch

Konfekt committed:
add case distinction for cygwin

Konfekt committed:
escape for # and % in :Open needed despite shellescape(..., 1)

Konfekt committed:
show error on missing standard handler

as suggested in [0]

___

[0]: https://github.com/vim/vim/pull/15721#issuecomment-2439727123

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,86 @@ set cpo&vim
+        \ exe 'silent !'.. (&shell =~? '\<cmd\.exe\>' ? '' : 'cmd.exe /c')
+        \ 'start /b ' trim(<q-args>) s:redir | redraw!
+endif
+if exists(':Launch') == 2
```
This can be rewritten as:

```vim
if exists(':Launch') == 2
  function s:open(file)
    " Default / Git Bash
    let cmd = ''
    " Windows / WSL
    if executable('explorer.exe')
      let cmd = 'explorer.exe'
    " Linux / BSD
    elseif executable('xdg-open')
      let cmd = 'xdg-open'
    " MacOS
    elseif executable('open')
      let cmd = 'open'
    endif

    if empty(cmd) && !exists('g:netrw_browsex_viewer')
      echoerr "No program found to open this file. For more information see :help Open."
    else
      Launch cmd shellescape(a:file, 1)
    endif
  endfunction
  command -complete=file -nargs=1 Open call s:open(<q-args>)
endif
```

Konfekt replied:
Yeah, I thought about it but found a check on every function call out of proportion for the unlikely event that one of these missing handlers without must be installed while Vim cannot be restarted

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,86 @@ set cpo&vim
+    if empty(a:cmd) && !exists('g:netrw_browsex_viewer')
+      echoerr "No executable handler found. See :help Open for default handlers. To customize it, see :help netrw_browsex_viewer"
+    else
+      Launch cmd shellescape(file, 1)
```
```suggestion
      Launch cmd shellescape(a:file, 1)
```

Konfekt replied:
Oops, thanks a ton

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,86 @@ set cpo&vim
+      Launch cmd shellescape(file, 1)
+    endif
+  endfunction
+  command -complete=file -nargs=1 Open call s:Open(<q-args>, s:cmd)
```
```suggestion
  command -complete=file -nargs=1 Open call s:Open(s:cmd, <q-args>)
```

Konfekt replied:
Oh my, I must have been rather absent-minded, thank you

saccarosium commented on runtime/plugin/netrwPlugin.vim:
```diff
@@ -31,6 +32,86 @@ set cpo&vim
+  endif
+  function s:Open(cmd, file)
+    if empty(a:cmd) && !exists('g:netrw_browsex_viewer')
+      echoerr "No executable handler found. See :help Open for default handlers. To customize it, see :help netrw_browsex_viewer"
```
How about "No program to open this file found. See :help Open for more information."? I think using 'executable handler' is a bit confusing.

Konfekt replied:
I openend for path (instead of say, locator) as a common term for file and URLs

Konfekt committed:
fix typo

Co-authored-by: Luca Saccarola <96259932+saccarosium@users.noreply.github.com>

Konfekt committed:
fix mixup

Co-authored-by: Luca Saccarola <96259932+saccarosium@users.noreply.github.com>

Konfekt committed:
follow @saccarosium 's suggestion

cpplearner commented on runtime/doc/pi_netrw.txt:
```diff
@@ -1469,42 +1469,93 @@ With either form of the command, netrw will first ask for confirmation
+Netrw determines which special handler by the following method:
+
+  * if |g:netrw_browsex_viewer| exists, then it will be used to attempt to
+    view files.  Examples of useful settings (place into your <.vimrc>):
```
```suggestion
    view files.
```
